### PR TITLE
chore: Batch update dev dependencies

### PR DESCRIPTION
**Pull request by aahung**
### Description of changes

ruff adds a new rule that catches these:
```
bin/parse_docs.py:95:13: PLW2901 Outer for loop variable `description` overwritten by inner assignment target
bin/parse_docs.py:96:13: PLW2901 Outer for loop variable `description` overwritten by inner assignment target
bin/parse_docs.py:102:13: PLW2901 Outer for loop variable `description` overwritten by inner assignment target
samtranslator/model/api/api_generator.py:524:17: PLW2901 Outer for loop variable `basepath` overwritten by inner assignment target
samtranslator/model/api/http_api_generator.py:405:17: PLW2901 Outer for loop variable `path` overwritten by inner assignment target
samtranslator/open_api/open_api.py:222:21: PLW2901 Outer for loop variable `param` overwritten by inner assignment target
samtranslator/third_party/py27hash/hash.py:81:13: PLW2901 Outer for loop variable `y` overwritten by inner assignment target
samtranslator/translator/translator.py:331:21: PLW2901 Outer for loop variable `connector_logical_id` overwritten by inner assignment target
samtranslator/utils/py27hash_fix.py:412:17: PLW2901 Outer for loop variable `arg` overwritten by inner assignment target
```

### Description of how you validated changes

`make pr` works

I tried to run a single integration test to see if pytest dependencies have breaking changes. Turns out fine:

```
(samtranslator37) ~/P/serverless-application-model ❯❯❯ pytest --no-cov integration/single/test_basic_function.py -k test_function_with_http_api_events                                                        ✘ 1 
=============================================================================================== test session starts ===============================================================================================
platform darwin -- Python 3.7.2, pytest-7.2.1, pluggy-0.13.1
rootdir: /Users/xinhol/Projects/serverless-application-model, configfile: pytest.ini
plugins: env-0.8.1, xdist-3.2.0, rerunfailures-11.1.1, cov-4.0.0, forked-1.3.0
collected 21 items / 19 deselected / 2 selected                                                                                                                                                                   

integration/single/test_basic_function.py::TestBasicFunction::test_function_with_http_api_events_0_single_function_with_http_api_events 
------------------------------------------------------------------------------------------------- live log setup --------------------------------------------------------------------------------------------------
INFO     botocore.credentials:credentials.py:1313 Found credentials in shared credentials file: ~/.aws/credentials
PASSED                                                                                                                                                                                                      [ 50%]
integration/single/test_basic_function.py::TestBasicFunction::test_function_with_http_api_events_1_single_function_alias_with_http_api_events 
-------------------------------------------------------------------------------------------------- live log call --------------------------------------------------------------------------------------------------
WARNING  samtranslator.metrics.metrics:metrics.py:135 There are unpublished metrics. Please make sure you call publish after you record all metrics.
PASSED                                                                                                                                                                                                      [100%]

================================================================================== 2 passed, 19 deselected in 196.64s (0:03:16) ===================================================================================
```

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: requirements/dev.txt [hunk 1]
```diff
@@ -1,22 +1,21 @@
...
-pytest~=6.2.5
-parameterized~=0.7.4
+pytest~=7.2
+parameterized~=0.8.1
 
 # Integration tests
-dateparser~=0.7
+dateparser~=1.1
 boto3>=1.23,<2
-tenacity~=7.0.0
```
Comment (aahung) [inline requirements/dev.txt:16]: note: this is not used.

**Review by ssenchenko — APPROVED** [review]

**Review by acristin — APPROVED** [review]

**Review by xazhao — APPROVED** [review]

File: samtranslator/model/api/http_api_generator.py [hunk 2]
```diff
@@ -401,15 +401,13 @@ def _construct_basepath_mappings(
                 if re.search(invalid_regex, path) is not None:
                     raise InvalidResourceException(self.logical_id, "Invalid Basepath name provided.")
 
-                # ignore leading and trailing `/` in the path name
-                path = path.strip("/")
```
Comment (hoffa) [inline samtranslator/model/api/http_api_generator.py:405]: This has no impact on `path` on L404?

File: samtranslator/third_party/py27hash/hash.py [hunk 3]
```diff
@@ -78,8 +78,7 @@ def thash(value):  # type: ignore[no-untyped-def]
         for y in value:
             length -= 1
 
-            y = Hash.hash(y)
-            x = (x ^ y) * mult
+            x = (x ^ Hash.hash(y)) * mult
```
Comment (hoffa) [inline samtranslator/third_party/py27hash/hash.py:81]: What did it complain about? Reusing `y`? 2 assignments was clearer to me.

**Review by hoffa — APPROVED** [review]

File: samtranslator/third_party/py27hash/hash.py (hunk 3 above)
Comment (aahung) [reply samtranslator/third_party/py27hash/hash.py:81]: See description

File: samtranslator/model/api/http_api_generator.py (hunk 2 above)
Comment (aahung) [reply samtranslator/model/api/http_api_generator.py:405]: I moved 405 and 404 should follow

**Pull request merged**